modules/packages/UnrolledLinkedList.chpl

```diff
@@ -0,0 +1,1249 @@
+      while cur != nil {
+        if idx >= pos + cur!.size {
+          pos += cur!.size;
+          cur = cur!.next;
+        }
+        else {
+          // The position in this node
+          var i = idx - pos;
+          
+          // The node to insert into
+          var dest = cur;
+
+          if cur!.size == nodeCapacity {
+            // The current node is full
+            try! {
+              var newNode = _split(cur!);
+              if i >= cur!.size {
+                dest = newNode;
+                i -= cur!.size;
+              }
+            }
+          }
+
+          // Shift elements by one step
+          dest!.size += 1;
+          for j in i+1..dest!.size-1 by -1 {
+            dest!.data[j] = dest!.data[j-1];
+          } 
+
+          dest!.data[i] = x;
+
+          break;
+        }
+      }
+    }
+
+    /*
+      Insert an element at a given position in this unrolledLinkedList, shifting all elements
+      currently at and following that index one to the right. The call
+      ``a.insert(0, x)`` inserts an element at the front of the unrolledLinkedList `a`, and
+      ``a.insert((a.size), x)`` is equivalent to ``a.append(x)``.
+
+      If the insertion is successful, this method returns `true`. If the given
+      index is out of bounds, this method does nothing and returns `false`.
+
+      .. warning::
+      
+        Inserting an element into this unrolledLinkedList may invalidate existing references
+        to the elements contained in this unrolledLinkedList.
+
+      :arg idx: The index into this unrolledLinkedList at which to insert.
+      :type idx: `int`
+
+      :arg x: The element to insert.
+      :type x: `eltType`
+
+      :return: `true` if `x` was inserted, `false` otherwise.
+      :rtype: `bool`
+    */
+    proc ref insert(idx: int, x: eltType): bool
+         lifetime this < x {
+      var result = false;
+
+      on this {
+        _enter();
+
+        // Handle special case of `a.insert((a.size), x)` here.
+        if idx == _size {
+          _append(x);
+          result = true;
+        } else if _withinBounds(idx) {
+          _insert(idx, x);
+          result = true;
+        }
+
+        _leave();
+      }
+
+      return result;  
+    }
+
+    pragma "no doc"
+    proc ref _insertGenericKnownSize(idx: int, items, size: int): bool {
+      //TODO: Maybe some optimization here. This O(N^2)
+      var result = false;
+
+      _sanity(size >= 0);
+
+      if size == 0 then
+        return true;
+
+
+      on this {
+        if idx == _size {
+          for x in items {
+            _append(x);
+          }
+          result = true;
+        }
+        else if _withinBounds(idx) {
+          var count = 0;
+          for x in items {
+            _insert(idx + count, x);
+            count += 1;
+          }
+          result = true;
+        }
+      }
+
+      return result;
+    }
+
+    /*
+      Insert an array of elements `arr` into this unrolledLinkedList at index `idx`,
```

Comment:
I would like to reword this. A novice Chapel programmer might assume that we are actually inserting an array. Though what we are doing is inserting all the elements in it.